cohere/compass/utils.py

```
@@ -46,11 +47,18 @@ def imap_queued(
             done, futures_set = futures.wait(
                 futures_set, return_when=futures.FIRST_COMPLETED
             )
+
             for future in done:
-                yield future.result()
+                try:
+                    yield future.result()
+                except Exception as e:
+                    logger.exception(f"Error in processing file: {e}")
```

I personally don't particularly like swallowing exceptions at this level, since it makes error handling more difficult and ends up relying just on the logs to know that an error has happened. `imap_queued` is a low level method, so ideally it should either 1) fail entirely and throw an exception and expects the user to retry 2) fail partially and return the list of exceptions to the user.

---

So the problem right now is:
* We send 10 files to parser api
* SDK divides them into 2 jobs, 5 files each on two parser nodes A and B
* Parser A restarts on file 3, not because of a failure on file but due to k8s
* SDK fails the entire job (even tho parser B is still processing files)

This fix is about retrying this use-case

> 1) fail entirely and throw an exception and expects the user to retry

I don't think this is right, according to the use case above

> 2) fail partially and return the list of exceptions to the user.

Why should a user retry for the above explained use case? Please note that I have seen the above use case quite a few times. This was the best solution I could think of

I hope this explanation helps in clarification :) Feel free to book a f2f meeting if needed.

---

Thanks, Ankush, for the explanation. Resolving the conversation.